### PR TITLE
TP-489: Plugin to perform validation on annotations 

### PR DESCRIPTION
**Pull request by maheshsubramanian**
… in maven 'proces sources' lifecycle phase

### CODE REVIEW

File: annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/domain/ValidationResult.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    public void setClassWithAnnotation(final String classWithAnnotation) {
+        this.classWithAnnotation = classWithAnnotation;
+    }
+
+    public boolean isValidationPassed() {
+        return validationPassed;
+    }
+
+    public void setValidationPassed(boolean validationPassed) {
```
Comment (mapingo) [inline annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/domain/ValidationResult.java]: validationPassed could be final

File: annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/domain/ValidationResult.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        this.validationPassed = validationPassed;
+    }
+
+    public String getValidationText() {
+        return validationText;
+    }
+
+    public void setValidationText(String validationText) {
```
Comment (mapingo) [inline annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/domain/ValidationResult.java]: final

File: annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojo.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    private List<ValidationResult> processAnnotation(final URLClassLoader classLoader, final String annotationClass) throws MojoFailureException {
+
+        List<ValidationResult> annotationValidationResults = newArrayList();
+        final ClassInfoList classesWithAnnotation = new ClassGraph().enableAnnotationInfo().scan().getClassesWithAnnotation(annotationClass);
+        AnnotationValidator annotationValidator = getValidator(annotationClass);
+        getLog().info("Found " + classesWithAnnotation.getNames().size() + " classes for annotation '" + annotationClass + "'");
+        for (String noa : classesWithAnnotation.getNames()) {
+            try {
+                annotationValidationResults.add(annotationValidator.validate(forName(noa, false, classLoader), of("serviceName", serviceName)));
+            } catch (ClassNotFoundException e) {
```
Comment (mapingo) [inline annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojo.java]: ClassNotFoundException could be final

File: annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojo.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+            reportError("Found " + failedValidationCount + "validation errors");
+        } else {
+            getLog().info("No validation errors found.");
+        }
+
+    }
+
+    private List<ValidationResult> processAnnotation(final URLClassLoader classLoader, final String annotationClass) throws MojoFailureException {
+
+        List<ValidationResult> annotationValidationResults = newArrayList();
```
Comment (mapingo) [inline annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojo.java]: final

File: annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojo.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    private void initialiseClassLoader() throws MojoFailureException {
+        try {
+            if (null == classLoader) {
+                final String targetClassesDirectory = project.getBuild().getOutputDirectory();
+                getLog().info("Consuming classes from directory : " + targetClassesDirectory);
+
+                classLoader = new URLClassLoader(new URL[]{new File(targetClassesDirectory).toURI().toURL()}, currentThread().getContextClassLoader());
+            }
+            currentThread().setContextClassLoader(classLoader);
+        } catch (MalformedURLException e) {
```
Comment (mapingo) [inline annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojo.java]: MalformedURLException could be final

File: annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojo.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+            getLog().info("No validation errors found.");
+        }
+
+    }
+
+    private List<ValidationResult> processAnnotation(final URLClassLoader classLoader, final String annotationClass) throws MojoFailureException {
+
+        List<ValidationResult> annotationValidationResults = newArrayList();
+        final ClassInfoList classesWithAnnotation = new ClassGraph().enableAnnotationInfo().scan().getClassesWithAnnotation(annotationClass);
+        AnnotationValidator annotationValidator = getValidator(annotationClass);
```
Comment (mapingo) [inline annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojo.java]: final

File: annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/report/AnnotationValidationReport.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+
+public class AnnotationValidationReport extends AbstractValidationReport {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AnnotationValidationReport.class);
+
+    @Override
+    protected String getFileContents(final ReportConfig reportConfig) {
+
+        List<String> csvRows = newArrayList();
```
Comment (mapingo) [inline annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/report/AnnotationValidationReport.java]: final

File: annotation-validator-maven-plugin/src/test/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojoTest.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    @After
+    public void tearDown() {
+        for (File f : new File(PROJECT_BASE_DIR).listFiles()) {
+            if (f.getName().startsWith(PLUGIN_VALIDATION_RESULT_PREFIX)) {
+                f.delete();
+            }
+        }
+    }
+
+    public void testShouldScanClassesWithAnnotationsAndGenerateReportWithoutFailingBuild() throws Exception {
```
Comment (mapingo) [inline annotation-validator-maven-plugin/src/test/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojoTest.java]: normally we start the test method names with "should"

File: annotation-validator-maven-plugin/src/test/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojoTest.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        final File pom = new File(PROJECT_BASE_DIR + "/pom-generate-report.xml");
+        final AnnotationCheckMojo mojo = (AnnotationCheckMojo) lookupConfiguredMojo(pom, "check-annotations");
+
+        mojo.execute();
+
+        // assert report generation and content
+        verifyReport();
+    }
+
+    public void testShouldScanClassesWithAnnotationsAndGenerateReportWhilstFailingBuild() throws Exception {
```
Comment (mapingo) [inline annotation-validator-maven-plugin/src/test/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojoTest.java]: as above test method name should start with "should" :-)

File: annotation-validator-maven-plugin/src/test/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojoTest.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+            fail("Should have thrown an exception to fail build");
+        } catch (MojoFailureException e) {
+
+            // assert report generation and content
+            verifyReport();
+        }
+
+    }
+
+    public void testShouldSkipValidation() throws Exception {
```
Comment (mapingo) [inline annotation-validator-maven-plugin/src/test/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojoTest.java]: start with should

File: annotation-validator-maven-plugin/src/test/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojoTest.java (hunk 8 above)
Comment (maheshsubramanian) [reply annotation-validator-maven-plugin/src/test/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojoTest.java]: need the test prefix as these MoJO tests are not using the @Test annotation

File: annotation-validator-maven-plugin/src/test/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojoTest.java (hunk 9 above)
Comment (maheshsubramanian) [reply annotation-validator-maven-plugin/src/test/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojoTest.java]: @mapingo same as above

File: annotation-validator-maven-plugin/src/test/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojoTest.java (hunk 10 above)
Comment (maheshsubramanian) [reply annotation-validator-maven-plugin/src/test/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojoTest.java]: @mapingo same as above

File: annotation-validator-maven-plugin/pom.xml [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+        </dependency>
+        <dependency>
+            <groupId>io.github.classgraph</groupId>
+            <artifactId>classgraph</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>uk.gov.justice.framework-api</groupId>
+            <artifactId>framework-api-domain</artifactId>
+            <version>2.2.0</version>
+            <scope>compile</scope>
```
[the remark this answers is not in the document]
Comment (matt-rich) [reply annotation-validator-maven-plugin/pom.xml]: Also we should be using a property for the version

File: annotation-validator-maven-plugin/pom.xml [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+        </dependency>
+        <dependency>
+            <groupId>uk.gov.justice.maven</groupId>
+            <artifactId>maven-test-utils</artifactId>
+            <version>${project.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.github.classgraph</groupId>
+            <artifactId>classgraph</artifactId>
```
[the remark this answers is not in the document]
Comment (maheshsubramanian) [reply annotation-validator-maven-plugin/pom.xml]: done

File: annotation-validator-maven-plugin/pom.xml (hunk 11 above)
[the remark this answers is not in the document]
Comment (maheshsubramanian) [reply annotation-validator-maven-plugin/pom.xml]: done

File: annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/domain/ValidationResult.java (hunk 1 above)
Comment (maheshsubramanian) [reply annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/domain/ValidationResult.java]: done

File: annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/domain/ValidationResult.java (hunk 2 above)
Comment (maheshsubramanian) [reply annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/domain/ValidationResult.java]: done

File: annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojo.java (hunk 6 above)
Comment (maheshsubramanian) [reply annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/annotations/AnnotationCheckMojo.java]: done

File: annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/report/AnnotationValidationReport.java (hunk 7 above)
Comment (maheshsubramanian) [reply annotation-validator-maven-plugin/src/main/java/uk/gov/justice/plugin/report/AnnotationValidationReport.java]: done

**Comment by maheshsubramanian** [comment]
Not required any more...

**Pull request closed without merging**